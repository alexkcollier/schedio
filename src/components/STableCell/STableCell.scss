.table__cell {
  padding: spacing('base');

  &--numeric {
    text-align: right;
  }
}
