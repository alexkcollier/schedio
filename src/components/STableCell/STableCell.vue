<template>
  <component
    :is="cellType"
    v-bind="$attrs"
    :class="{ 'table__cell--numeric': isNumeric }"
    class="table__cell"
  >
    <slot />
  </component>
</template>

<script>
export default {
  name: 'STableCell',

  props: {
    isHeader: {
      type: Boolean,
      default: false
    },
    isNumeric: {
      type: Boolean,
      default: false
    }
  },

  data() {
    return {
      parentTag: ''
    }
  },

  computed: {
    cellType() {
      return this.isInHeaderRow ? 'th' : 'td'
    },
    isInHeaderRow() {
      return this.isHeader || this.parentTag === 'THEAD'
    }
  },

  mounted() {
    // make $el available to computed
    this.parentTag = this.$parent.$el.parentNode.tagName
  },

  updated() {
    if (this.isNumeric && isFinite(this.$el.innerText)) {
      this.$el.innerText = Number(this.$el.innerText).toLocaleString()
    }
  }
}
</script>
