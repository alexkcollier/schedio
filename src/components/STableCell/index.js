import STableCell from './STableCell.vue'

export { STableCell }

export default { install: Vue => Vue.component(STableCell.name, STableCell) }
