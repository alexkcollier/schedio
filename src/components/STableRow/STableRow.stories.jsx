import { SHeading } from '@/components/SHeading'
import { STable } from '@/components/STable'
import { STableCell } from '@/components/STableCell'
import { STableRow } from '@/components/STableRow'
import PropList from '@@/docs/components/PropList'
import { text } from '@storybook/addon-knobs'

export default {
  title: 'Components|Tables/STableRow',

  parameters: {
    jest: 'STableRow',
  },
}

export const tableRow = () => ({
  props: {
    slotContent: {
      default: text('default', 'Slot content', 'Slots'),
    },
  },
  render (h) {
    return (
      <div>
        <SHeading level="1">Table Row</SHeading>
        <p>
          Table rows are containers for <code>{'<STableCell />'}</code>. They do not have any
            props, but have a single default slot. <code>{'<STableCell />'}</code> is the only
          content that should go in this slot.
        </p>

        <SHeading level="2">Example</SHeading>
        <STable>
          <STableRow slot="header">
            <STableCell>Header</STableCell>
          </STableRow>

          <STableRow>{this.slotContent}</STableRow>
        </STable>

        {STableRow.props && <PropList component={STableRow} />}
      </div>
    )
  },
})

tableRow.story = {
  name: 'Table Row',
}
