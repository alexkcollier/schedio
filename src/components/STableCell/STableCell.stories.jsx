import { SHeading } from '@/components/SHeading'
import { STable } from '@/components/STable'
import { STableCell } from '@/components/STableCell'
import { STableRow } from '@/components/STableRow'
import PropList from '@/docs/PropList'
import { boolean, number, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/Tables.STableCell', module)
  .addParameters({ jest: 'STableCell' })
  .add('STableCell', function STableCellDocs() {
    return {
      props: {
        isNumeric: { default: boolean('is-numeric', true, 'Optional Props') },
        isHeader: { default: boolean('is-header', false, 'Optional Props') },
        headerCellSlotContent: { default: text('Header content', 'Header cell slot', 'Slots') },
        numericCellSlotContent: { default: number('Numeric content', 1200, {}, 'Slots') }
      },
      render(h) {
        return (
          <div>
            <SHeading level="1">Table Cell</SHeading>
            <p>
              Table cells contain the actual data values for the table. They should always be
              enclosed in <code>{'<STableRow />'}</code>.
            </p>
            <p>
              Cells placed within a <code>{'<thead />'}</code> will automatically use the{' '}
              <code>{'<th />'}</code> tag instead of <code>{'<td />'}</code>.
            </p>
            <p>
              If <code>isNumeric</code> is <code>true</code>, the cell will try to convert to its
              localized version with <code>Number.toLocaleString()</code>
            </p>

            <SHeading level="2">Example</SHeading>
            <STable>
              <STableRow slot="header">
                <STableCell>
                  Toggling <code style="text-transform: none;">isHeader</code> will change this one
                </STableCell>
                <STableCell>
                  Toggling <code style="text-transform: none;">isNumeric</code> will change this one
                </STableCell>
              </STableRow>

              <STableRow>
                <STableCell is-header={this.isHeader}>{this.headerCellSlotContent}</STableCell>
                <STableCell is-numeric={this.isNumeric}>{this.numericCellSlotContent}</STableCell>
              </STableRow>
            </STable>

            <PropList component={STableCell} />
          </div>
        )
      }
    }
  })
