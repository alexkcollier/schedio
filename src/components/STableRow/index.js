import STableRow from './STableRow.vue';

export { STableRow };

export default { install: Vue => Vue.component(STableRow.name, STableRow) };
