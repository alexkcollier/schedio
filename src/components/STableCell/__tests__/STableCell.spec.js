import { mount } from '@vue/test-utils'
import STableCell from '@/components/STableCell/STableCell.vue'
import STableRow from '@/components/STableRow/STableRow.vue'

describe('STableCell.vue', () => {
  let wrapper
  let tableCell

  beforeEach(() => {
    wrapper = mount({
      name: 'BaseWrapper',
      components: { STableCell, STableRow },
      template: `
        <tbody>
          <STableRow>
            <STableCell>Content</STableCell>
          </STableRow>
        </tbody>
      `,
    })

    tableCell = wrapper.find(STableCell)
  })

  it('renders correctly', () => {
    expect(tableCell.html()).toMatchSnapshot()
  })

  // mount this before test or `STableRow` tag will not update
  const headerWrapper = mount(
    {
      name: 'HeaderWrapper',
      components: { STableCell, STableRow },
      template: `
        <thead>
          <STableRow>
            <STableCell>Content</STableCell>
          </STableRow>
        </thead>
      `,
    },
    {
      attachToDocument: true,
    },
  )

  it("automatically detects if it's a header", () => {
    expect(wrapper.contains('th')).toBe(false)
    expect(headerWrapper.find(STableCell).html()).toMatchSnapshot()
    expect(headerWrapper.contains('th')).toBe(true)
  })

  it('can be manually set to use `th`', async () => {
    expect(wrapper.contains('th')).toBe(false)
    tableCell.setProps({ isHeader: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.contains('th')).toBe(true)
    expect(tableCell.html()).toMatchSnapshot()
  })

  it('can be numeric', async () => {
    tableCell.setProps({ isNumeric: true })
    await wrapper.vm.$nextTick()
    expect(tableCell.html()).toMatchSnapshot()
  })
})
