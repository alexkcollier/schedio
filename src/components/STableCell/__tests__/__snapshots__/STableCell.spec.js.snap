// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`STableCell.vue automatically detects if it's a header 1`] = `<th class="table__cell">Content</th>`;

exports[`STableCell.vue can be manually set to use \`th\` 1`] = `<th class="table__cell">Content</th>`;

exports[`STableCell.vue can be numeric 1`] = `<td class="table__cell table__cell--numeric">Content</td>`;

exports[`STableCell.vue renders correctly 1`] = `<td class="table__cell">Content</td>`;
